### PR TITLE
Bedrock support

### PR DESCRIPTION
**Pull request by CreeperG16**
This PR adds Bedrock Edition item support to the Item class.
Todo:
- [x] ~~Implement Stack ID in Item class~~ This should be moved to prismarine-registry later
- [x] Add support for older bedrock versions, which might have different formats
    - Add more features to `features.json` for bedrock in minecraft-data, this way testing if the registry is bedrock could be skipped entirely in some cases -> [PrismarineJS/minecraft-data#696](https://github.com/PrismarineJS/minecraft-data/pull/696)
- [x] ~~Items with different namespaces - mostly affects block names in CanPlaceOn and CanDestroy~~ this can be addressed later if there are problems
- [ ] anvil - minor, could do later?
- [x] toNotch and fromNotch functions
    - [x] < 1.16.220
    - [x] >= 1.16.220
    - [x] canPlaceOn and canDestroy
    - [x] stack ID
    - [x] blockingTick
- [x] Docs
- [x] Types
- [ ] Tests
    - [x] Handle Stack ID in PC tests
        - [x] ~~Temporary solution: setting the same stack ID~~
        - [x] Better solution: Stack ID is null if the version doesn't support Stack IDs (right now just test if bedrock, but could be added as an mcdata feature maybe?)
        - [ ] LT solution: use Item.equal() in tests - need to implement NBT equal function to work properly
    - [x] Bedrock tests

### CODE REVIEW

**Comment by CreeperG16** [comment]
```
    // TODO:
    // - tests
    // - docs
    // - add support for older versions:
    //   need to see how items are handled in older bedrock protocol versions,
    //   and probably update features.json in minecraft-data with useful data
    // - Stack ID generation
    // - Block runtime ID field
    // - Setting canPlaceOn and canDestroy with block/item objects?
    // - Merge with Item class in ../index.js?```
I'm probably forgetting some things

**Comment by CreeperG16** [comment]
Sorry, just realized that I missed #85 completely
Would it be better to close this and continue #85? Or would that complicate things even more? Sorry, I'm still a bit new to GitHub

File: tools/relay.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+process.env.DEBUG = process.argv.includes('--debug') && 'minecraft-protocol'
```
Comment (extremeheat) [inline tools/relay.js:1]: remove

File: package.json [hunk 2]
```diff
@@ -16,6 +16,7 @@
   },
   "devDependencies": {
     "@types/node": "^18.6.4",
+    "bedrock-protocol": "^3.25.0",
```
Comment (extremeheat) [inline package.json:19]: should not be needed, mc-data or [mc-packets](https://github.com/PrismarineJS/minecraft-packets) should be used instead

File: lib/bedrock-item.js [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    canBePlacedOnBlock (block) {
+      if (typeof block === 'string') {
+        return this.canPlaceOn.includes(block) || this.canPlaceOn.includes(`minecraft:${block}`)
+      } else {
+        return this.canDestroy.includes(block.name) || this.canPlaceOn.includes(`minecraft:${block.name}`)
+      }
+    }
+
+    /** @returns {boolean} */
+    canDestroyBlock (block) {
```
Comment (extremeheat) [inline lib/bedrock-item.js:195]: Needs to be added to pc implementation also

File: lib/bedrock-item.js [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+      return this.nbt?.value?.Damage?.value ?? 0
+    }
+
+    set durabilityUsed (value) {
+      if (!this.nbt) this.nbt = { name: '', type: 'compound', value: {} }
+      this.nbt.value.Damage = { type: 'int', value }
+    }
+
+    get spawnEggMobName () {
+      return this.name.replace('_spawn_egg', '')
```
Comment (extremeheat) [inline lib/bedrock-item.js:182]: is this accurate?

File: lib/bedrock-item.js [hunk 5]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    /** @returns {boolean} */
+    canBePlacedOnBlock (block) {
```
Comment (extremeheat) [inline lib/bedrock-item.js:186]: canBePlacedOnBlock -> canPlaceOn

File: lib/bedrock-item.js [hunk 6]
```diff
@@ -0,0 +1,207 @@
...
+    }
+
+    // Stack ID
+    static currentStackId = 0
+    static nextStackId () {
+      return BedrockItem.currentStackId++
+    }
+
+    toNetwork (serverAuthoritative = true) {
+      if (Object.keys(this).length === 0) return { network_id: 0 }
```
Comment (extremeheat) [inline lib/bedrock-item.js:87]: Replace this check with `this.type == 0`

File: lib/bedrock-item.js [hunk 7]
```diff
@@ -0,0 +1,207 @@
...
+      } else if (item2 == null) {
+        return false
+      } else {
+        return (
+          item1.type === item2.type &&
+                    (matchStackSize ? item1.count === item2.count : true) &&
+                    item1.metadata === item2.metadata &&
+                    JSON.stringify(item1.nbt) === JSON.stringify(item2.nbt) &&
+                    (sameStack ? item1.stackId === item2.stackId : true) &&
+                    (placeable ? item1.canPlaceOn.sort().toString() === item2.canPlaceOn.sort().toString() : true) &&
```
Comment (extremeheat) [inline lib/bedrock-item.js:74]: too complicated, can this logic be simplified?

Comment (CreeperG16) [reply lib/bedrock-item.js:74]: Should it not be optional to check if it can be placed on / destroy the same blocks? or maybe just not check that at all?

File: lib/bedrock-item.js (hunk 5 above)
Comment (CreeperG16) [reply lib/bedrock-item.js:186]: canPlaceOnBlock? the field with the array is called canPlaceOn

Comment (extremeheat) [reply lib/bedrock-item.js:186]: Then those fields need to be private or be called something else

File: lib/bedrock-item.js (hunk 4 above)
Comment (CreeperG16) [reply lib/bedrock-item.js:182]: As far as I have tested on the current version. Probably not on older versions, but I would need to test more for that (spawneggs is mostly why I added it to the todo list)

File: lib/bedrock-item.js (hunk 7 above)
Comment (extremeheat) [reply lib/bedrock-item.js:74]: This can just be considered an extended part of what would normally be NBT data. You can add a `checkNbt = true` parameter and check against nbt / the bedrock `canPlaceOn` things

File: tools/relay.js [hunk 8]
```diff
@@ -0,0 +1,31 @@
+const { Relay } = require('bedrock-protocol')
```
Comment (extremeheat) [inline tools/relay.js:1]: Remove file

**Review by extremeheat — COMMENTED** [review]
Remember the API needs to be the same as the pc implementation

File: lib/bedrock-item.js (hunk 5 above)
Comment (CreeperG16) [reply lib/bedrock-item.js:186]: if private then I should add methods to add/remove blocks it can be placed on/destroy
or would something like canPlaceOnArray and canDestroyArray work? I can't think of a better name

File: tools/relay.js (hunk 8 above)
Comment (CreeperG16) [reply tools/relay.js:1]: ah, just thought it would be useful to have in the environment

File: lib/bedrock-item.js (hunk 5 above)
Comment (extremeheat) [reply lib/bedrock-item.js]: If these is these fields are exposed, this would need to be added to the pc version also (`canPlaceOn()`, `canDestoryBlock()`). This includes the stackId field. There shouldn't need to be version checks on basic prismarine-item functions (except for the protocol import/export functions) as it's meant to abstract the item, not just re-expose protocol specific internals. 

In the future, they might add these functions to the NBT. So for now adding `breakableBlocks` and `placeableBlocks` field to prismarine-item may make sense, and on bedrock these things refer to the special fields sent over the network.

File: index.js [hunk 9]
```diff
@@ -204,6 +224,37 @@ function loader (registryOrVersion) {
       }
       throw new Error("Don't know how to get spawn egg mob name for this mc version")
     }
+
+    // TODO: test this
+    canPlaceOn (block) {
+      if (!this.nbt || !this.nbt.value.CanPlaceOn) return true
```
Comment (extremeheat) [inline index.js:230]: Is this actually a thing?

File: index.js [hunk 10]
```diff
@@ -32,45 +34,51 @@ function loader (registryOrVersion) {
...
-        return notchItem
+        if (item.nbt && item.nbt.length !== 0) {
+          networkItem.nbtData = item.nbt
+        }
+        return networkItem
       }
       throw new Error("Don't know how to serialize for this mc version ")
     }
 
-    static fromNotch (item) {
```
Comment (extremeheat) [inline index.js:73]: Revert , breaking change

File: index.js (hunk 9 above)
Comment (CreeperG16) [reply index.js:230]: according to the wiki it is, but like I said I need to test it

Comment (extremeheat) [reply index.js:230]: Link?

Comment (CreeperG16) [reply index.js:230]: also probably not a thing on every version

Comment (CreeperG16) [reply index.js:231]: https://minecraft.fandom.com/wiki/Tutorials/Command_NBT_tags#Items
now that I look back, CanPlaceOn isn't there, only CanDestroy

Comment (CreeperG16) [reply index.js:231]: https://minecraft.fandom.com/wiki/Tutorials/Command_NBT_tags#Blocks
It only exists for blocks, which makes sense as only blocks can be placed

Comment (extremeheat) [reply index.js:231]: Ok, then if it's like this for Java Edition it is most likely stored this way on Bedrock Edition also on disk. In which case the canPlaceOn/canDestroy should not be fields on Item instance. They should retrieve and store data in NBT like enchantments.

**Comment by extremeheat** [comment]
Since everything seems mostly identical to pc, with the exception of import/export functions, one class seems to make sense if enchantments are the same (minus integer vs string IDs). Are they roughly the same?

**Comment by CreeperG16** [comment]
other than the stack ID stuff, yes, they seem to be pretty much the same

File: index.js (hunk 9 above)
Comment (CreeperG16) [reply index.js:231]: So handle them the same way enchantments are handled?

File: lib/bedrock-item.js (hunk 4 above)
Comment (CreeperG16) [reply lib/bedrock-item.js:183]: The mob name is in the spawn egg item's name from 1.16.100, according to [the wiki](https://minecraft.fandom.com/wiki/Spawn_Egg#History). Probably could add this as the `spawnEggsHaveSpawnedEntityInName` feature in minecraft-data

**Comment by CreeperG16** [comment]
What I've noticed is a lot of things could already work with the java code if the same features were there in bedrock's features.json in minecraft-data. The spawn egg mob names for example, the getter would need no modification if the features were in minecraft-data

**Comment by extremeheat** [comment]
> bedrock's features.json in minecraft-data. The spawn egg mob names for example, the getter would need no modification if the features were in minecraft-data

Right, you can open a PR to add it to the features

File: index.js [hunk 11]
```diff
@@ -140,37 +189,98 @@ function loader (registryOrVersion) {
...
+        }
+      }
+    }
+
+    // TODO: namespaces other than minecraft:
+    // The item palette on bedrock can contain items
+    // with other namespaces, probably need to handle
+    // this in prismarine-registry
+    // Also this needs to be tested
+    get canPlaceOn () {
```
Comment (extremeheat) [inline index.js:248]: canPlaceOn -> blocksCanPlaceOn
canDestroy -> blocksCanDestroy

Also must be added to doc that this applies to adventure mode.
```js
this.nbt.value.CanPlaceOn = nbt.list(nbt.string( blocks.map(b => b.startsWith('minecraft:') ? b : 'minecraft:' + b ) ))
```

**Comment by extremeheat** [comment]
Relevant line in minecraft java edition code for the block break/place rules - https://github.dev/extremeheat/extracted_minecraft_data/blob/4b469221a8a4559e634823d931dcae26d59f548e/client/net/minecraft/world/item/ItemStack.java#L931

File: index.js [hunk 12]
```diff
@@ -140,37 +196,96 @@ function loader (registryOrVersion) {
...
 
-      // The 'registry.itemsByName[this.name].maxDurability' checks to see if this item can lose durability
-      if (registry.supportFeature('whereDurabilityIsSerialized') === 'Damage' && registry.itemsByName[this.name].maxDurability) {
-        this.nbt.value.Damage = { type: 'int', value: 0 }
+        // The 'registry.itemsByName[this.name].maxDurability' checks to see if this item can lose durability
+        if (
+          registry.supportFeature('whereDurabilityIsSerialized') === 'Damage' &&
+          registry.itemsByName[this.name].maxDurability
+        ) {
+          this.nbt.value.Damage = { type: 'int', value: 0 }
```
Comment (CreeperG16) [inline index.js:245]: Why exactly is Damage set in the enchantment setter?

File: index.js [hunk 13]
```diff
@@ -140,37 +196,96 @@ function loader (registryOrVersion) {
...
       throw new Error("Don't know how to get the enchants from an item on this mc version")
     }
 
     set enchants (normalizedEnchArray) {
-      const isBook = this.name === 'enchanted_book'
-      const enchListName = registry.supportFeature('nbtNameForEnchant')
-      const type = registry.supportFeature('typeOfValueForEnchantLevel')
-      if (type === null) throw new Error("Don't know the serialized type for enchant level")
-      if (!this.nbt) this.nbt = { name: '', type: 'compound', value: {} }
+      if (registry.type === 'bedrock') {
```
Comment (extremeheat) [inline index.js:208]: Make a feature?

File: index.js [hunk 14]
```diff
@@ -140,37 +196,96 @@ function loader (registryOrVersion) {
...
+          this.nbt.value.Damage = { type: 'int', value: 0 }
+        }
       }
     }
 
+    // TODO: namespaces other than minecraft:
+    // The item palette on bedrock can contain items
+    // with other namespaces, probably need to handle
+    // this in prismarine-registry
+    // Also this needs to be tested
```
Comment (extremeheat) [inline index.js:254]: We don't handle non-minecraft namespaces on Java Edition either, so there is no need to worry about it for bedrock. Our priority is vanilla for now, but if big problems arise it can be investigated later

File: index.js [hunk 15]
```diff
@@ -140,37 +196,96 @@ function loader (registryOrVersion) {
...
+    get blocksCanPlaceOn () {
+      if (Object.keys(this).length === 0) return []
+
+      if (!this.nbt?.value?.CanPlaceOn) return []
+
+      return nbt.simplify(this.nbt).CanPlaceOn
+    }
+
+    set blocksCanPlaceOn (blocks) {
+      if (!this.nbt) this.nbt = { name: '', type: 'compound', value: {} }
```
Comment (extremeheat) [inline index.js:264]:
```js
this.nbt = nbt.comp({})
```

The nbt builder functions can be used elsewhere also for better readability

File: index.js [hunk 16]
```diff
@@ -140,37 +196,96 @@ function loader (registryOrVersion) {
...
+      this.nbt.value.CanDestroy = nbt.list(
+        nbt.string(blocks.map((b) => (b.startsWith('minecraft:') ? b : `minecraft:${b}`)))
+      )
+    }
+
     get durabilityUsed () {
       if (Object.keys(this).length === 0) return null
       const where = registry.supportFeature('whereDurabilityIsSerialized')
-      if (where === 'Damage') {
+      if (where === 'Damage' || registry.type === 'bedrock') {
```
Comment (extremeheat) [inline index.js:288]: Can be put into feature

File: index.js [hunk 17]
```diff
@@ -32,105 +33,188 @@ function loader (registryOrVersion) {
...
       } else if (registry.supportFeature('itemSerializationUsesBlockId')) {
-        if (item.blockId === -1) return null
-        return new Item(item.blockId, item.itemCount, item.itemDamage, item.nbtData)
+        if (networkItem.blockId === -1) return null
+        return new Item(networkItem.blockId, networkItem.itemCount, networkItem.itemDamage, networkItem.nbtData)
+      } else if (registry.type === 'bedrock') {
+        if (registry.version['<']('1.16.220')) {
+          // unsure about this, different packets use slightly different formats, but everything
+          // in the item field stays the same - just sometimes it isn't in the item field
+          const item = new Item(
```
Comment (extremeheat) [inline index.js:134]: Can you inline these for clarity?

File: index.d.ts [hunk 18]
```diff
@@ -5,38 +5,72 @@ import { Tags, TagType } from 'prismarine-nbt'
...
+    } & ({ has_stack_id: 0 } | { has_stack_id: 1; stack_id: number }))
+  | ({
+      // < 1.16.220
+      network_id: number;
+      auxiliary_value: number;
+      can_place_on: string[];
+      can_destroy: string[];
+    } & ({ has_nbt: false } | { has_nbt: true; nbt: { version: 1; nbt: Tags[TagType] } }));
+
+declare type NetworkItem = PcNetworkItem | BedrockNetworkItem;
```
Comment (extremeheat) [inline index.d.ts:73]: unnecessary

File: index.js [hunk 19]
```diff
@@ -32,105 +33,188 @@ function loader (registryOrVersion) {
...
+            metadata: item.metadata,
+            has_stack_id: +serverAuthoritative,
+            stack_id: serverAuthoritative ? item.stackId : undefined,
+            block_runtime_id: 0, // TODO
+            extra: {
+              has_nbt: item.nbt !== null,
+              nbt: item.nbt !== null ? { version: 1, nbt: item.nbt } : undefined,
+              can_place_on: item.canPlaceOn,
+              can_destroy: item.canDestroy
+              // blocking_tick: 0, // TODO
```
Comment (extremeheat) [inline index.js:110]: Set to 0

File: index.d.ts [hunk 20]
```diff
@@ -5,38 +5,72 @@ import { Tags, TagType } from 'prismarine-nbt'
...
+  slot: number;
+  count: number;
+  metadata: number;
+  nbt: Tags[TagType] | null;
+  stackId: number;
+  name: string;
+  displayName: string;
+  stackSize: number;
+  durabilityUsed: number;
+  enchants: NormalizedEnchant[];
```
Comment (extremeheat) [inline index.d.ts:19]:
```ts
get enchants(): { name: string, lvl: number }[]
set enchants(enchantments: { name: string, lvl: number }[])
```

File: index.js [hunk 21]
```diff
@@ -32,105 +33,188 @@ function loader (registryOrVersion) {
...
       }
       throw new Error("Don't know how to deserialize for this mc version ")
     }
 
     get customName () {
-      if (Object.keys(this).length === 0) return null
-      return this?.nbt?.value?.display?.value?.Name?.value ?? null
+      if (Object.keys(this).length === 0 || !this.nbt) return null
+
+      return nbt.simplify(this.nbt).display?.Name ?? null
```
Comment (extremeheat) [inline index.js:163]: This get function can be simplified to one line (the other checks are not necessary with optional chanining):
```
return this.nbt?.value?.display?.value?.Name?.value
```

File: index.js [hunk 22]
```diff
@@ -32,105 +33,188 @@ function loader (registryOrVersion) {
...
+      }
     }
 
     // gets the cost based on previous anvil uses
     get repairCost () {
-      if (Object.keys(this).length === 0) return 0
-      return this?.nbt?.value?.RepairCost?.value ?? 0
+      if (Object.keys(this).length === 0 || !this.nbt) return 0
+
+      return nbt.simplify(this.nbt).RepairCost ?? 0
```
Comment (extremeheat) [inline index.js:193]: Same as above comment

File: index.js [hunk 23]
```diff
@@ -140,37 +224,88 @@ function loader (registryOrVersion) {
...
+        nbt.string(blocks.map((b) => (b.startsWith('minecraft:') ? b : `minecraft:${b}`)))
+      )
+    }
+
+    get blocksCanDestroy () {
+      if (Object.keys(this).length === 0) return []
+
+      if (!this.nbt?.value?.CanDestroy) return []
+
+      return nbt.simplify(this.nbt).CanDestroy
```
Comment (extremeheat) [inline index.js:295]: Same as above comment

File: index.js (hunk 19 above)
Comment (CreeperG16) [reply index.js:110]: If I understand the [docs](https://minecraft-data.prismarine.js.org/?v=bedrock_1.19.62&d=protocol#Item) correctly it's present if the item is the shield and absent if it isn't? That should be easy enough to test for

File: index.d.ts [hunk 24]
```diff
@@ -5,42 +5,73 @@ import { Tags, TagType } from 'prismarine-nbt'
...
+    blocking_tick?: number;
+  };
+
+  // < 1.16.220
+  auxiliary_value?: number;
+  can_place_on?: string[];
+  can_destroy?: string[];
 }
 
+declare type NetworkItem = PcNetworkItem | BedrockNetworkItem;
```
Comment (extremeheat) [inline index.d.ts:75]: remove

File: index.js (hunk 19 above)
Comment (extremeheat) [reply index.js:110]: It should be set to 0 at all times, the serializer will take care of writing it or not

File: index.js [hunk 25]
```diff
@@ -202,6 +323,13 @@ function loader (registryOrVersion) {
       if (registry.supportFeature('spawnEggsHaveSpawnedEntityInName')) {
         return this.name.replace('_spawn_egg', '')
       }
+      if (registry.type === 'bedrock') {
+        // TODO: add to features.json
+        // According to the wiki, the different metadata values were split
+        // into their own IDs in 1.16.100
+        // https://minecraft.fandom.com/wiki/Spawn_Egg#History
```
Comment (extremeheat) [inline index.js:330]: Since spawn eggs don't change much, indeed if it's necessary a manual map of item name to entity name may make sense

Comment (CreeperG16) [reply index.js:330]: I don't think that's necessary, as far as I can tell, the mob's name is in the item's name, and in older versions it uses the entity's internalId in metadata
metadata is still supported in the current version (in commands such as /give) but I think the spawn egg names are used in protocol, I'll look into that

**Comment by CreeperG16** [comment]
Setting stack ID in tests is temporary, need to implement an NBT equal function as JSON.stringify() doesn't work if fields are in a different order

File: index.d.ts [hunk 26]
```diff
@@ -5,42 +5,75 @@ import { Tags, TagType } from 'prismarine-nbt'
...
 }
 
-declare interface NotchItem {
-    // 1.8 - 1.12
-    blockId?: number;
-    itemDamage?: number;
-    // 1.13 - 1.15
-    present?: boolean;
-    itemId?: number;
+declare interface PcNetworkItem {
```
Comment (extremeheat) [inline index.d.ts:40]: Remove

File: index.d.ts [hunk 27]
```diff
@@ -5,42 +5,75 @@ import { Tags, TagType } from 'prismarine-nbt'
...
-    itemCount?: number;
-    nbtData?: Buffer;
+  itemCount?: number;
+  nbtData?: Buffer;
 }
 
-declare interface NormalizedEnchant {
-    name: string;
-    lvl: number
+declare interface BedrockNetworkItem {
```
Comment (extremeheat) [inline index.d.ts:52]: Remove

File: test/anvil.test.js [hunk 28]
```diff
@@ -11,6 +11,8 @@ describe('1.8.9 anvil', () => {
     const inverse = Item.anvil(sword2, sword1, false, undefined)
     const finalItem = Item.fromNotch({ blockId: 276, itemCount: 1, itemDamage: 0, nbtData: { type: 'compound', name: '', value: { ench: { type: 'list', value: { type: 'compound', value: [{ lvl: { type: 'short', value: 1 }, id: { type: 'short', value: 34 } }] } }, RepairCost: { type: 'int', value: 3 } } } })
     const inverseFinalItem = Item.fromNotch({ blockId: 276, itemCount: 1, itemDamage: 0, nbtData: { type: 'compound', name: '', value: { ench: { type: 'list', value: { type: 'compound', value: [{ lvl: { type: 'short', value: 1 }, id: { type: 'short', value: 34 } }] } }, RepairCost: { type: 'int', value: 3 } } } })
+    finalItem.stackId = res.item.stackId
```
Comment (extremeheat) [inline test/anvil.test.js:14]: Put these into the static constructor

File: test/anvil.test.js [hunk 29]
```diff
@@ -36,6 +40,7 @@ describe('1.8.9 anvil', () => {
     const inverse = Item.anvil(book, sword, false, undefined)
     const finalItem = Item.fromNotch({ blockId: 276, itemCount: 1, itemDamage: 3, nbtData: { type: 'compound', name: '', value: { ench: { type: 'list', value: { type: 'compound', value: [{ lvl: { type: 'short', value: 5 }, id: { type: 'short', value: 16 } }] } }, RepairCost: { type: 'int', value: 3 } } } })
     const inverseFinalItem = null
+    finalItem.stackId = res.item.stackId
```
Comment (extremeheat) [inline test/anvil.test.js:43]: Same as above

File: test/basic.test.js [hunk 30]
```diff
@@ -28,19 +28,19 @@ describe('test based on examples', () => {
     const ironShovelItem = new Item(472, 1)
 
     it('constructor makes item correctly', () => {
-      const expectedObj = { count: 1, displayName: 'Iron Shovel', metadata: 0, name: 'iron_shovel', nbt: null, stackSize: 1, type: 472 }
+      const expectedObj = { count: 1, displayName: 'Iron Shovel', metadata: 0, name: 'iron_shovel', nbt: { name: '', type: 'compound', value: { Damage: { type: 'int', value: 0 } } }, stackSize: 1, type: 472, stackId: 0 }
```
Comment (extremeheat) [inline test/basic.test.js:31]: Why the change with nbt?

Comment (CreeperG16) [reply test/basic.test.js:31]: As in vanilla as far as I have tested on both java and bedrock, any item which can lose durability has the Damage tag set to 0 by default, so I added that to the constructor; then I had to change it here as the iron shovel item will have the Damage tag in NBT

File: index.d.ts (hunk 27 above)
Comment (CreeperG16) [reply index.d.ts:52]: The entire interface?

File: test/anvil.test.js (hunk 28 above)
Comment (CreeperG16) [reply test/anvil.test.js:14]: How would I assign the stack ID from the static functions? Or just set it to 0?

File: index.d.ts (hunk 27 above)
Comment (extremeheat) [reply index.d.ts:52]: Yes, types should hold stable API things and network packets are not stable in structure (eg new fields may be added)

File: test/anvil.test.js (hunk 28 above)
Comment (extremeheat) [reply test/anvil.test.js:14]: See your API doc. Second parameter

Comment (CreeperG16) [reply test/anvil.test.js:14]: Ah, forgot that I did that. I should add it to the types too

File: index.js [hunk 31]
```diff
@@ -123,21 +123,22 @@ function loader (registryOrVersion) {
     static fromNotch (networkItem, stackId) {
       if (registry.supportFeature('itemSerializationWillOnlyUsePresent')) {
         if (networkItem.present === false) return null
-        return new Item(networkItem.itemId, networkItem.itemCount, networkItem.nbtData)
+        return new Item(networkItem.itemId, networkItem.itemCount, networkItem.nbtData, stackId)
```
Comment (extremeheat) [inline index.js:126]: why?

File: index.js [hunk 32]
```diff
@@ -123,21 +123,22 @@ function loader (registryOrVersion) {
...
       } else if (registry.supportFeature('itemSerializationAllowsPresent')) {
         if (networkItem.itemId === -1 || networkItem.present === false) return null
-        return new Item(networkItem.itemId, networkItem.itemCount, networkItem.nbtData)
+        return new Item(networkItem.itemId, networkItem.itemCount, networkItem.nbtData, stackId)
```
Comment (extremeheat) [inline index.js:129]: why?

File: index.js [hunk 33]
```diff
@@ -123,21 +123,22 @@ function loader (registryOrVersion) {
...
       } else if (registry.type === 'bedrock') {
+        if (networkItem.network_id === 0) return new Item(0)
         if (registry.supportFeature('itemSerializeUsesAuxValue')) {
           const item = new Item(networkItem.network_id, networkItem.auxiliary_value & 0xff, networkItem.auxiliary_value >> 8, networkItem.nbt?.nbt, stackId)
           item.blocksCanPlaceOn = networkItem.can_place_on
           item.blocksCanDestroy = networkItem.can_destroy
           return item
         } else {
-          const item = new Item(networkItem.network_id, networkItem.count, networkItem.metadata, networkItem.extra.nbt?.nbt, networkItem.stack_id)
+          const item = new Item(networkItem.network_id, networkItem.count, networkItem.metadata, networkItem.extra.nbt?.nbt, networkItem.stack_id ?? stackId)
```
Comment (extremeheat) [inline index.js:141]: why?

**Review by extremeheat — COMMENTED** [review]
Remove all the unnecessary stackId defining and `??` operations, including from the tests. The only time it should be specified in the tests is making sure there is equality. and if so it should be set to `null`.

**Comment by extremeheat** [comment]
What's the status of this? Stale PRs are not good as they will accumulate merge conflicts over time

**Comment by CreeperG16** [comment]
sorry, haven't had much time the past few weeks but should have some now... The prismarine-registry PR is pretty much done (just need to do some testing) so there's just this left. I might be able to finish it this weekend but I'll see what I need to fix

**Comment by CreeperG16** [comment]
Only anvil left, hopefully anvil usage doesn't differ very much from java to bedrock.

**Comment by rom1504** [comment]
I think it would be nice to get this in

1. how do we check it works for your purpose in bedrock ?
2. how do we check it doesn't break pc ?

anvil support is minor

File: index.js [hunk 34]
```diff
@@ -15,122 +16,203 @@ function loader (registryOrVersion) {
...
-    static toNotch (item) {
+    // Stack ID
+    // Probably needs to be moved to prismarine-registry later on as calling the loader again will reset it
+    static currentStackId = 0
+    static nextStackId () {
+      return Item.currentStackId++
+    }
+
+    static toNotch (item, serverAuthoritative = true) {
       if (registry.supportFeature('itemSerializationAllowsPresent')) {
```
Comment (extremeheat) [inline index.js:69]: This function is hard to read as you do a feature check for registry type on bedrock but not pc.  You can simplify this function with 

```js
static toNotch(item, serverAuthoritative = true) {
  const hasNBT = item.nbt && item.nbt.value.length > 0
  if (registry.type === 'pc') {
    if (registry.supportFeature('itemSerializationAllowsPresent')) {
      if (item == null) return { present: false }
      return {
        present: true,
        itemId: item.type,
        itemCount: item.count,
        nbtData: item.nbt
      }
    } else if (registry.supportFeature('itemSerializationUsesBlockId')) {
      if (item == null) return { blockId: -1 }
      return {
        blockId: item.type,
        itemCount: item.count,
        itemDamage: item.metadata,
        nbtData: item.nbt
      }
    }
  } else if (registry.type === 'bedrock') {
    if (item == null || item.type === 0) return { network_id: 0 }
    if (registry.supportFeature('itemSerializeUsesAuxValue')) {
      return {
        network_id: item.id,
        auxiliary_value: (item.metadata << 8) | (item.count & 0xff),
        can_place_on: item.blocksCanPlaceOn,
        can_destroy: item.blocksCanDestroy,
        blocking_tick: 0,
        has_nbt: hasNBT,
        nbt: hasNBT ? { version: 1, nbt: item.nbt } : undefined
      }
    } else {
      return {
        network_id: item.type,
        count: item.count,
        metadata: item.metadata,
        has_stack_id: serverAuthoritative,
        stack_id: serverAuthoritative ? item.stackId : undefined,
        block_runtime_id: 0,
        extra: {
          can_place_on: item.blocksCanPlaceOn,
          can_destroy: item.blocksCanDestroy,
          blocking_tick: 0,
          has_nbt: hasNBT,
          nbt: hasNBT ? { version: 1, nbt: item.nbt } : undefined
        }
      }
    }
  }
  throw new Error("Don't know how to serialize for this mc version ")
}
```

**Review by extremeheat — COMMENTED** [review]
LGTM, left some comments on the current code. For the anvil handling code, maybe we can do that in a separate PR and leave the Item.anvil methods as undefined for bedrock for now

File: index.js [hunk 35]
```diff
@@ -66,48 +66,42 @@ function loader (registryOrVersion) {
...
+
+      if (registry.type === 'pc') {
+        if (registry.supportFeature('itemSerializationAllowsPresent')) {
+          if (item == null) return { present: false }
+          const networkItem = {
+            present: true,
+            itemId: item.type,
+            itemCount: item.count
+          }
+          if (hasNBT) networkItem.nbtData = item.nbt
```
Comment (extremeheat) [inline index.js:79]: this should not be necessary, it will just be set as undefined if assigned to networkItem. You can just do 

```
nbtData: hasNBT ? item.nbt : undefined
```

inside the object.

Comment (CreeperG16) [reply index.js:79]: I set it because it still sets `nbtData` to undefined, and that caused the tests to fail because they're expecting the field not to be there. I'll change the tests then.

File: README.md [hunk 36]
```diff
@@ -82,7 +90,21 @@ the item's custom lore (ie. set in give command)
...
+
+Returns an array of enchants on the Item with their name and level
+
+#### set item.enchants({ name: string, lvl: number }[])
+
+Updates the Item's NBT enchantments based on assigned array
+
+#### item.blocksCanPlaceOn
+
+In adventure mode, the list of block names (as strings) that this Item can be placed on
```
[the remark this answers is not in the document]
Comment (extremeheat) [reply README.md:103]: Returning a prismarine-block isn't what I meant. I meant returning properties assosiated with a block. Like `stone[variant=andersite]` --> `[["stone", {"variant":"andersite"]]` as opposed to just `["stone"]`.

[the remark this answers is not in the document]
Comment (CreeperG16) [reply README.md:103]: ah, ok

[the remark this answers is not in the document]
Comment (CreeperG16) [reply README.md:103]: Not sure how to type/document it, if it never returns those properties anyway as there isn't a way of getting them, should they be documented to return those properties at all?

**Comment by rom1504** [comment]
Hey, does it work ?

Does it break PC? Will it help for bedrock?

**Comment by CreeperG16** [comment]
It works as far as I've tested, and I've written bedrock tests for it too. PC tests are also working fine, but I haven't tried PC much outside of that, though I made sure to keep the original functionality intact. As for helping with bedrock, apart from anvil, it implements all methods and properties PC has.

**Review by rom1504 — APPROVED** [review]
